src/bsoncxx/types/bson_value/value.cpp

```diff
@@ -25,6 +27,170 @@ BSONCXX_INLINE_NAMESPACE_BEGIN
 namespace types {
 namespace bson_value {
 
+value::value(b_array v) : value(v.value) {}
+value::value(b_binary v) : value(v.type_id, v.sub_type, v.size, v.bytes) {}
+value::value(b_bool v) : value(v.value) {}
+value::value(b_code v) : value(v.type_id, v) {}
+value::value(b_codewscope v) : value(type::k_codewscope, v.code, v.scope) {}
+value::value(b_date v) : value(v.value) {}
+value::value(b_dbpointer v) : value(v.type_id, v.collection, v.value) {}
+value::value(b_decimal128 v) : value(v.value) {}
+value::value(b_document v) : value(v.view()) {}
+value::value(b_double v) : value(v.value) {}
+value::value(b_int32 v) : value(v.value) {}
+value::value(b_int64 v) : value(v.value) {}
+value::value(b_maxkey) : value(type::k_maxkey) {}
+value::value(b_minkey) : value(type::k_minkey) {}
+value::value(b_null) : value(nullptr) {}
+value::value(b_oid v) : value(v.value) {}
+value::value(b_regex v) : value(v.type_id, std::string{v.regex}, std::string{v.options}) {}
+value::value(b_symbol v) : value(v.type_id, v) {}
+value::value(b_timestamp v) : value(v.type_id, v.increment, v.timestamp) {}
+value::value(b_undefined) : value(type::k_undefined) {}
+value::value(b_utf8 v) : value(v.value) {}
+
+value::value(decimal128 v) : value(type::k_decimal128, v.high(), v.low()) {}
+
+value::value(const type id) : _impl{stdx::make_unique<impl>()} {
+    if (id == type::k_minkey) {
+        _impl->_value.value_type = BSON_TYPE_MINKEY;
+    } else if (id == type::k_maxkey) {
+        _impl->_value.value_type = BSON_TYPE_MAXKEY;
+    } else if (id == type::k_undefined) {
+        _impl->_value.value_type = BSON_TYPE_UNDEFINED;
+    } else {
+        throw std::logic_error{"Must be min/max key or undefined"};
```

Comment:
> There's also the open issue of how to handle errors. As of now I just throw an error but I'm also working on that too

Is the open question whether or not to throw? There is precedent for throwing in some constructors. At least client comes to mind: https://github.com/mongodb/mongo-cxx-driver/blob/78831317d0dde2b36c32236303991d3737c23972/src/mongocxx/client.cpp#L84

Reply:
Yeah, I guess I'll have to throw regardless, for the memcpy and allocations I think. I'll clean this up.